modules/core/src/main/java/org/apache/ignite/spi/discovery/tcp/ServerImpl.java

```diff
@@ -6139,6 +6144,13 @@ private void processMetricsUpdateMessage(TcpDiscoveryMetricsUpdateMessage msg) {
                         Map<Integer, CacheMetrics> cacheMetrics = msg.hasCacheMetrics(nodeId) ?
                             msg.cacheMetrics().get(nodeId) : Collections.emptyMap();
 
+                        if (metricsSizeNotProcessed && cacheMetrics.size() >= METRICS_QNT_WARN) {
+                            log.warning("The Discovery message has metrics for " + cacheMetrics.size() +
+                                " caches.\nTo prevent Discovery blocking use" +
+                                " -DIGNITE_DISCOVERY_DISABLE_CACHE=true option.");
```

Comment:
It's the same code as in ServerImpl - you'd better to move this in protected method to TcpDiscoveryImpl